File: ios/TennisAce/Features/Matches/MatchesViewModel.swift
import Foundation
import Observation

/// Holds the matches-list state and loads from the `MatchRepository`. The view
/// observes the published properties and only ever calls `load()` back.
@MainActor
@Observable
final class MatchesViewModel {
    private let repository: MatchRepository

    private(set) var isLoading = true
    private(set) var matches: [Match] = []
    private(set) var errorMessage: String?

    init(repository: MatchRepository) {
        self.repository = repository
    }

    func load() async {
        isLoading = true
        errorMessage = nil
        do {
            matches = try await repository.matches()
        } catch {
            errorMessage = error.localizedDescription
        }
        isLoading = false
    }
}

File: android/app/src/main/java/investments/hoelzl/tennisace/ui/matches/MatchesViewModel.kt
package investments.hoelzl.tennisace.ui.matches

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import investments.hoelzl.tennisace.data.MatchRepository
import investments.hoelzl.tennisace.model.Match
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/** Immutable UI state for the matches list. */
data class MatchesUiState(
    val isLoading: Boolean = true,
    val matches: List<Match> = emptyList(),
    val errorMessage: String? = null,
)

/**
 * Holds the matches-list state and loads from the [MatchRepository]. The UI
 * observes [uiState] and only ever sends intents back (e.g. [load]).
 */
class MatchesViewModel(
    private val repository: MatchRepository,
) : ViewModel() {

    private val _uiState = MutableStateFlow(MatchesUiState())
    val uiState: StateFlow<MatchesUiState> = _uiState.asStateFlow()

    init {
        load()
    }

    fun load() {
        _uiState.update { it.copy(isLoading = true, errorMessage = null) }
        viewModelScope.launch {
            try {
                val matches = repository.matches()
                _uiState.update { it.copy(isLoading = false, matches = matches) }
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(isLoading = false, errorMessage = e.message ?: "Unbekannter Fehler")
                }
            }
        }
    }

    companion object {
        /** Factory that injects the [repository] into the ViewModel. */
        fun factory(repository: MatchRepository): ViewModelProvider.Factory =
            object : ViewModelProvider.Factory {
                @Suppress("UNCHECKED_CAST")
                override fun <T : ViewModel> create(modelClass: Class<T>): T =
                    MatchesViewModel(repository) as T
            }
    }
}
